File: Data Structures & Algorithms/count-number-of-islands/submission-1.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows, cols = len(grid), len(grid[0])
        visit = set()
        q = deque()
        res = 0

        def bfs(r, c):
            q.append((r, c))
            visit.add((r, c))

            while q:
                r, c = q.popleft()
                dir = [[0, 1], [0, -1], [1, 0], [-1, 0]]

                for dx, dy in dir:
                    row = dx + r
                    col = dy + c

                    if row in range(rows) and col in range(cols) and grid[row][col] == "1" and (row, col) not in visit:
                        q.append((row, col))
                        visit.add((row, col))

        
        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == "1" and (row, col) not in visit:
                    bfs(row, col)
                    res += 1
        
        return res

File: Data Structures & Algorithms/set-zeroes-in-matrix/submission-1.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        // int m = matrix.size();
        // int n = matrix[0].size();
        // vector<bool> rows(m);
        // vector<bool> cols(n);

        // for (int r = 0; r < m; r++) {
        //     for (int c = 0; c < n; c++) {
        //         if (matrix[r][c] == 0) {
        //             rows[r] = true;
        //             cols[c] = true;
        //         }
        //     }
        // }

        // for (int r = 0; r < m; r++) {
        //     for (int c = 0; c < n; c++) {
        //         if (rows[r] || cols[c]) {
        //             matrix[r][c] = 0;
        //         }
        //     }
        // }

        int m = matrix.size();
        int n = matrix[0].size();

        bool firstRowZero = false;
        bool firstColZero = false;

        for (int c = 0; c < n; c++) {
            if (matrix[0][c] == 0) {
                firstRowZero = true;
            }
        }

        for (int r = 0; r < m; r++) {
            if (matrix[r][0] == 0) {
                firstColZero = true;
            }
        }

        for (int r = 1; r < m; r++) {
            for (int c = 1; c < n; c++) {
                if (matrix[r][c] == 0) {
                    matrix[r][0] = 0;
                    matrix[0][c] = 0;
                }
            }
        }

        for (int r = 1; r < m; r++) {
            for (int c = 1; c < n; c++) {
                if (
                    matrix[r][0] == 0 ||
                    matrix[0][c] == 0
                ) {
                    matrix[r][c] = 0;
                }
            }
        }

        if (firstRowZero) {
            for (int c = 0; c < n; c++) {
                matrix[0][c] = 0;
            }
        }

        if (firstColZero) {
            for (int r = 0; r < m; r++) {
                matrix[r][0] = 0;
            }
        }
    }
};